### DISCUSSION TITLE
Проверка наличия трека в своих плейлистах

### DISCUSSION THREAD

**Discussion by efremych** [Chimildic/goofy#110] [Личный опыт]
Ответы на вопросы [получил](https://github.com/Chimildic/goofy/discussions/109), спасибо @Chimildic, теперь можно и поделиться, что получилось.
Всё что требовалось для реализации, нашел тут, на форуме.

Используется [Аддон "Управление с телефона"](https://github.com/Chimildic/goofy/discussions/9).

1. В скрипте `launch.html` добавляю себе отдельную кнопку
```js
               {
                  action: 'checkCurPlaying',
                  title: 'Чек трека в плейлистах',
                  forms: [
                           { key: 'trackURI', placeholder: 'Ссылка на трек' },
                  ],
               },
```
![image](https://user-images.githubusercontent.com/76212904/126913145-7ca2dca2-e8a6-4592-a142-3392456a2aa0.png)

2. Что происходит по нажатию кнопки
- При вставке в текстовое поле ссылки на трек - в результате нам выводится информация о том, в каких личных плейлистах уже присутствует данный трек

- Если текстовое поле пусто то вызывается функция `RecentTracks.getPlayback()`. Т.е. текущий проигрываемый трек. И выводится всё та же информация
- Если же никакой трек не проигрывается в данный момент - видим соответствующее уведомление.

3. Вспомогательная шаблонная функция `Helper.parseId`. Вытаскивает нам из ссылки на трек id трека.

```js
const Helper = (function () {
  return {
    parseId: parseId,
  };

  function parseId(string) {
    let pattern = '[open.spotify.com|spotify]+[\/|:](track|playlist|album|artist|show|episode|concert|user)[\/|:]([^?/#& ]+)';
    let [fullMatch, type, id] = string.match(new RegExp(pattern, 'i')) || [];
    return id;
  }
})();
```

4. Непосредственно вызываемая из кнопки функция

```js
function checkCurPlaying(values) {
  let content = '';
  let playback = {};

  if (values.text.trackURI.length != 0) {
    playback.item = SpotifyRequest.getFullObjByIds('tracks', [Helper.parseId(values.text.trackURI)], 50)[0];
  } else {
    playback = RecentTracks.getPlayback();
      if (playback.isEmpty()) {
         content = 'Нет играющего трека и не вставлена ссылка на интересующий трек';
         return content;
      }
  }
    
  let playlists = [];
    
  Playlist.getPlaylistArray().forEach(p => {if (p.owner.id == User.id) {
    let tracks = Source.getPlaylistTracks('', p.id);
    let tracksMap = tracks.map((item) => item.id);
    if (tracksMap.includes(playback.item.id)) {
      playlists.push(tracks[0].origin.name);
    }
  }
 });

if (playlists.length == 0) {
  content = '\n' + '🎵' + `${playback.item.artists[0].name} - ${playback.item.name}` + '🎵'+ ' ОТСУТСТВУЕТ в плейлистах:';
} else {
  content = '\n' + '🎵' + `${playback.item.artists[0].name} - ${playback.item.name}` + '🎵'+ ' есть в плейлистах:' + '\n' + playlists.map((item) => item).join('***');
}

return content;
}
```
Итоговый результат выглядит примерно так:
![image](https://user-images.githubusercontent.com/76212904/126913853-b39d615c-c8a6-46ce-a061-d361d240aba7.png)

Тексты, естественно, можно поправить на свой вкус в коде. Как иметь возможность выводить результат с форматированием без потери '\n'  описано [здесь](https://github.com/Chimildic/goofy/discussions/109)
Можно еще заморочиться и исключать из итогового списка плейлистов - текущий проигрываемый плейлист (если это плейлист).
Вдруг, кому пригодится, или вдохновит на что-то.

**Comment by efremych** [comment]
Забыл, что изначально хотел не по id проверять трек, а по "Артист - Название трека".
Тогда блок выглядит так
```js
  Playlist.getPlaylistArray().forEach(p => {if (p.owner.id == User.getId()) {
    let tracks = Source.getPlaylistTracks('', p.id);
    let tracksMap = tracks.map((item) => `${item.artists[0].name} - ${item.name}`);
    if (tracksMap.includes(`${playback.item.artists[0].name} - ${playback.item.name}`)) {
      playlists.push(tracks[0].origin.name);
    }
  }
```

**Reply by Bedmit** [reply]
Я правильно понимаю, что проверяются только плейлисты, созданные самим пользователем? А как сделать, чтобы проверялись все, в том числе и зафолловленные плейлисты? 
Думал решить это с наскока заменой `owner` на `all` в `(p.owner.id == User.getId())` Все оказалось не так просто - выдает ошибку.

**Reply by efremych** [reply]
Чтобы проверялись все плейлисты, в том числе и зафоловленные, можно убрать условие на проверку id юзера. Но не знаю, насколько тогда будет тяжело скрипту (надо же из каждого треки вытащить), если много зафоловленных плейлистов. У меня-то всего 60 плейлистов (своих+зафоловленных)
```js
Playlist.getPlaylistArray().forEach(p => {
    let tracks = Source.getPlaylistTracks('', p.id);
    let tracksMap = tracks.map((item) => `${item.artists[0].name} - ${item.name}`.formatName());
    if (tracksMap.includes(`${playback.item.artists[0].name} - ${playback.item.name}`.formatName())) {
      playlists.push(tracks[0].origin.name);
    }
});
```
Насколько понимаю, так оно и работает. Если id userа мой, значит мой личный плейлист, все другие id - считается, что плейлист зафоловленный.

Вот так выглядит плейлист спотифая, например (зафоловлен у меня). На случай, если условия плодить
```js
  owner: 
   { display_name: 'Spotify',
     external_urls: { spotify: 'https://open.spotify.com/user/spotify' },
     href: 'https://api.spotify.com/v1/users/spotify',
     id: 'spotify',
     type: 'user',
     uri: 'spotify:user:spotify' }
```

**Reply by Bedmit** [reply]
Отлично. Естественно чуть дольше считает, но с моими плейлистами справляется. Правда пришлось еще одну открывающую фигурную скобку в первой строчке вставить.
Давно хотел какой-то инструмент отслеживания, из каких плейлистов треки попадают в мой плейлист дня. Спасибо!

**Reply by efremych** [reply]
> Правда пришлось еще одну открывающую фигурную скобку в первой строчке вставить

Криво скопировал, да и изначально, конечно, был не эталон форматирования :)
Проапдейтил в ответе блок кода от начала до точки с запятой. Могло не хватать чего-то в конце, или было что лишнее из-за кривого копирования, но в начале вроде никаких новых скобок не нужно.
